File: viking-vision.cpp
#include <iostream>
#include <ntcore.h>
#include <networktables/NetworkTable.h>
#include <opencv2/opencv.hpp>
#include <opencv2/xfeatures2d.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>
#include "utildefs.h"
#include "quad-filter.h"

using namespace cv;
using namespace cv::xfeatures2d;

// Grayscale
// Threshhold
// Blobbify
// Outline
// Identify quads / Image matching

bool parallelogram(std::vector<Point>)
{
	return true;
}

Mat toGrayscale(Mat src)
{
	Mat dst;
	cvtColor(src, dst, COLOR_BGR2GRAY);
	return dst;
}

Mat brightPass(Mat src)
{
	Mat dst;
	inRange(src, Scalar(230), Scalar(255), dst);
	return dst;
}

Mat blur(Mat src)
{
	Mat dst;
	GaussianBlur(src, dst, Size(13, 13), 40);
	return dst;
}

Mat blobbify(Mat src, int radius)
{
	Mat dst(src.rows, src.cols, CV_8U, Scalar(0));
	for (int r = 0; r < dst.cols; r++) //WHY DOES THIS WORK WTF TODO: understand this
		for (int c = 0; c < dst.rows; c++)
		{
			if (src.data[r + c * dst.cols] > 1)
			{
				ellipse(dst, Point(r, c), Size(radius, radius * 2 / 5), 90, 0, 360, Scalar(src.data[r + c * dst.cols]), -1);
			}
		}
	return dst;
}

Mat outline(Mat src)
{
	Mat dst(src.rows, src.cols, CV_8U, Scalar(0));
	std::vector<std::vector<Point>> contours, polys;
	std::vector<Point> poly;
	findContours(src, contours, CV_RETR_LIST, CV_CHAIN_APPROX_NONE);
	for (int i = 0; i < contours.size(); i++)
	{
		approxPolyDP(contours[i], poly, 5, true);
		polys.push_back(poly);
	}
	for (int i = 0; i < polys.size(); i++)
	{
		if (polys[i].size() == 4)
		{
			drawContours(dst, polys, i, Scalar(0xFF));
			circle(dst, polys[i][0], 3, Scalar(0xFF));
		}
	}
	dst = Mat::zeros(dst.rows, dst.cols, CV_8U);
	drawContours(dst, pairs(boundingBoxes(polys, (double)5/2, .75), 1), -1, Scalar(0xFF));
	return dst;
}

Mat quads(Mat src)
{
	return src;
}

int main(int argc, char** argv)
{
	VideoCapture cap;
	Mat frame;
	if (argc > 1 && !strcmp(argv[1], "-h"))
	{
		std::cout << "Usage: viking-vision <device>\ndevice defaults to 0" << std::endl;
		return 0;
	}

	if (argc > 1)
	{	
		cap = VideoCapture(atoi(argv[1]));
	} else
	{
		cap = VideoCapture(0);
	}

	if (!cap.isOpened())
	{
		std::cerr << "Failed to open device " << atoi(argv[1]) << std::endl;
		return -1;
	}
	namedWindow("Grayscale");
	namedWindow("Blur");
	namedWindow("BrightPass");
	namedWindow("Blobbed");
	namedWindow("Output", WINDOW_AUTOSIZE);
	for (int i = 0; i < 15; i++)
		cap >> frame; //Throw away first few frames
	while (true)
	{
		cap >> frame;
		frame = toGrayscale(frame);
		imshow("Grayscale", frame);
		frame = blur(frame);
		imshow("Blur", frame);
		frame = brightPass(frame);
		imshow("BrightPass", frame);
		frame = blobbify(frame, 15);
		imshow("Blobbed", frame);
		frame = outline(frame);

		imshow("Output", frame);
		if (waitKey(30) == 27)
		{
			return 0;
		}
	}
}

File: viking-vision.py
#byVisionDaddy
#!/usr/bin/env python2
from __future__ import division
import sys
import argparse
import cv2
import sys
import numpy as np
import logging
BLUR_SIZE = (11, 11)
BLUR_FACTOR = 90
GREEN_LOWER_LIMIT = 235
MEDIAN_BLUR_SIZE = 5
HEIGHT_WIDTH_RATIO = 5/2
HEIGHT_WIDTH_TOLERANCE = .85
AREA_TOLERANCE = 1.7
logging.basicConfig(level=logging.DEBUG)

ip = "10.29.28.56"

def ntInit(table):
    NetworkTables.initialize(ip)
    return NetworkTables.getTable(table)

# From http://stackoverflow.com/questions/4961017/clojure-style-function-threading-in-python
def T(*args):
    return reduce(lambda l, r: r(l), args)

def toGrayscale(src):
    return cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)

def toGreenscale(src):
    return cv2.extractChannel(src, 1)

def brightPass(src):
    return cv2.inRange(src, GREEN_LOWER_LIMIT, 255)

def blur(src):
    return cv2.GaussianBlur(src, BLUR_SIZE, BLUR_FACTOR)

def medianBlur(src):
    return cv2.medianBlur(src, MEDIAN_BLUR_SIZE)

def blobbify(src):
    data = src.flat
    dims = src.shape
    print dims
    for i in range(0, len(data)):
        if data[i] > 1:
            cv2.ellipse(src, (i % dims[0], i / dims[0]), (15, 6), 90, 0, 360, 0xFF, -1)
    return src

def outline(src):
    dst = np.zeros(src.shape, np.uint8)
    im2, contours, hierarchy  = cv2.findContours(src, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
    cv2.drawContours(dst, contours, -1, (255), 1)
    return dst, contours

def quads(contours):
    output = []
    boxes = []
    polys = [cv2.approxPolyDP(c, 5, True) for c in contours]
    polys = filter(lambda c: len(c) == 4, polys)
    for p in polys:
        _, _, w, h = cv2.boundingRect(p)
        boxes.append(abs(h/w - HEIGHT_WIDTH_RATIO) <= HEIGHT_WIDTH_TOLERANCE)
    for i in range(len(boxes)):
        if boxes[i]:
            output.append(polys[i])
    return output

def pairs(contours):
    areas = [cv2.contourArea(c) for c in contours]
    outputMask = [False] * len(areas)
    output = []
    # Awful imperative code TODO: write something better
    for i in range(len(areas)):
        for j in range(len(areas)):
            if i == j or outputMask[i] or outputMask[j]:
                continue
            if abs(max(areas[i], areas[j])/min(areas[i], areas[j])) - 1 <= AREA_TOLERANCE:
                outputMask[i] = True
                outputMask[j] = True
    for i in range(len(areas)):
        if outputMask[i]:
            output.append(contours[i])
    return output

def distanceToCenter(contours, frameWidth):
    moments = [cv2.moments(c) for c in contours][:2]
    normalized = []
    for m in moments:
        cx = int(m['m10'] / m['m00'])
        normalized.append(2*cx/frameWidth - 1)
    if len(normalized) > 0:
        return sum(normalized) / len(normalized)
    return -2

def degreesToCenter(contours, frameWidth):
    distances = []
    for m in [cv2.moments(c) for c in contours][:2]:
        cx = int(m['m10'] / m['m00'])
        distances.append(cx)
    if len(distances) > 0:
        distance = sum(distances) / len(distances)
        return (distance / frameWidth * 61) - 30.5
    return -2
        

def blobFilter(src):
    params = cv2.SimpleBlobDetector_Params()
    params.minThreshold = 0
    params.maxThreshold = 0xFF
    params.filterByArea = True
    params.minArea = 125
    params.maxArea = 10000
    params.filterByCircularity = True
    params.minCircularity = 0.5
    params.maxCircularity = 0.95
    params.filterByColor = False
    params.filterByConvexity = True
    params.minConvexity = 0.6
    params.maxConvexity = 1
    params.filterByInertia = True
    params.minInertiaRatio = 0
    params.maxInertiaRatio = 0.6
    params.minDistBetweenBlobs = 0
    detector = cv2.SimpleBlobDetector_create(params)
    keypoints = detector.detect(src)
    keypoints.sort(key = lambda kp: kp.size)
    keypoints = keypoints[:2]
    print keypoints
    return keypoints

def main(camera, display, haveNetworktables):
    if haveNetworktables:
        vc = ntInit('VisionControl')
    if display:
        cv2.namedWindow("Output")
    cap = cv2.VideoCapture(camera)
    ret, frame = cap.read()
    print frame.shape[0]
    distance, oldDistance = 0, 0
    while ret:
        distanceSent = False
        ret, frame = cap.read()
        frame = T(frame, toGreenscale, brightPass, medianBlur)
        '''frame, contours = outline(frame)
        contours = pairs(quads(contours))
        distance = distanceToCenter(contours, frame.shape[1])'''
        keypoints = blobFilter(frame)
        oldDistance = distance
        if len(keypoints) >= 1:
            distance = 61 * sum([kp.pt[0] for kp in keypoints]) / frame.shape[1] - 30.5
            oldDistance = distance * .65 + oldDistance * .35
        else:
            distance = oldDistance
        if abs(distance) <= 1:
            if haveNetworktables:
                vc.putValue("detectedValue", distance)
            distanceSent = True
        logging.info(str((distance, distanceSent)))
        if haveNetworktables:
            vc.putBoolean("targetLocked", distanceSent)
        if display:
            cv2.drawKeypoints(frame, keypoints, frame, 0x7F, cv2.DRAW_MATCHES_FLAGS_DRAW_OVER_OUTIMG)
            cv2.imshow("Output", frame)
        if(cv2.waitKey(30) & 0xFF == ord('q')):
            break
    cap.release()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Vision tracking script for FRC team 2928.")
    parser.add_argument("camera", help="camera input to use", type=int, default=0, nargs="?")
    parser.add_argument("-d", "--debug", help="displays the processed video feed (requires X Windows)", action="store_true")
    parser.add_argument("-n", "--no-networktables", help="disables networktables", action="store_true")
    args = parser.parse_args()
    if not args.no_networktables:
        from networktables import NetworkTables
    main(args.camera, args.debug, not args.no_networktables)
